File: agent/correlation.py
"""The one id that names a question in the trace, as it is named everywhere else.

The app mints it before the request leaves the browser and sends it as
`custom_inputs.request_id`; `player-insights-agent/shared/correlation.ts` is the
other half of this contract and holds the reasoning for why the browser mints it
and why it is not the run ledger's primary key. This module's job is narrower:
put that id where MLflow can be SEARCHED by it, so a question the app logged can
be turned into the trace of the run that answered it.

WHY TAGS AND NOT ONLY SPAN ATTRIBUTES. An attribute is readable once you have
already found the span, which is the problem, not the solution: `search_traces`
filters on tags. A correlation id that only exists as an attribute joins nothing
-- an operator holding an id from a log line still has to open traces one at a
time until one matches. So the ids are set as trace tags, and repeated as
attributes on the orchestrator span for a reader who opens it directly.

WHAT THIS SIDE KNOWS AND THE APP SIDE DOES NOT. The release facts here are the
ones baked into the model artifact or resolved by the served entity: the build
the agent was logged from, and the warehouse and spaces its data calls bill
against. The app records the other half beside them in the run ledger -- release
id, bundle target, app name, workspace and serving endpoint (`releaseIdentity` in
server/lib/run-admission.ts). Neither side guesses the other's: an agent reading
`DATABRICKS_APP_NAME` out of a serving container would record whichever app was
deployed last, or nothing, and both read as fact.

NOTHING CALLER-SUPPLIED IS PRINTED OR STORED UNVALIDATED. `custom_inputs` is an
untrusted body. A trace tag is queryable text stored beside a run and a print is
a log line, so an unchecked value there can forge a line, or carry a question's
text into the one place that deliberately does not record questions. `usable`
holds both ids to the shape the app mints and drops anything else, which is why
`request_id` reaches a tag through this module rather than directly.
"""

from __future__ import annotations

import re
from typing import Any

#: What the app prefixes its ids with, so ours are distinguishable in a log line
#: that also carries platform ids. Kept identical to `CORRELATION_PREFIX` in
#: shared/correlation.ts.
CORRELATION_PREFIX = "req-"

#: The prefix, then a hyphenated lowercase UUID, and nothing else. The same
#: expression as `CORRELATION_SHAPE` on the app side, and strict for the same
#: reason: this value is printed and stored.
_SHAPE = re.compile(r"^req-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")

#: The tag an operator searches by. Bare rather than dotted, because it is what
#: goes in `mlflow.search_traces(filter_string="tags.correlation_id = '...'")`
#: and a dotted key has to be backquoted there by everyone who ever uses it.
CORRELATION_TAG = "correlation_id"
#: The run ledger's primary key, so a Lakebase row leads to the trace and back.
#: Separate from the correlation id on purpose -- see shared/correlation.ts.
RUN_TAG = "run_id"


def usable(value: Any) -> str:
    """The caller's id if it is one we will print, otherwise the empty string.

    Empty is ordinary rather than exceptional: the benchmark runner, a `curl`,
    and any app build older than this contract all send nothing. A request is
    never refused over it, because a question's answer does not depend on
    whether its id was well formed.
    """

    if not isinstance(value, str):
        return ""
    trimmed = value.strip()
    return trimmed if _SHAPE.match(trimmed) else ""


def facts(required: Any, settings: Any = None) -> dict[str, str]:
    """The correlation and release facts for this turn, as strings.

    Absent values are OMITTED rather than written as empty strings. A tag whose
    value is "" reads as a measurement that came back blank; no tag at all reads
    as a version that did not record it, which is the truth for a request that
    sent no id and for a build logged before this existed.

    `required` is an `execution_identity.Requirement`, taken rather than the raw
    `custom_inputs` so there is one parser for the app's contract on this side of
    the wire. `settings` is the agent's `config.Settings`, and is optional
    because the two ids are worth recording even where settings are not to hand.
    """

    recorded: dict[str, str] = {}

    correlation = usable(getattr(required, "request_id", ""))
    if correlation:
        recorded[CORRELATION_TAG] = correlation
    run_id = usable(getattr(required, "run_id", ""))
    if run_id:
        recorded[RUN_TAG] = run_id

    # Which build of the agent answered. Baked at log time, so it is a property
    # of the served model version rather than of the container it runs in.
    build_sha = _text(getattr(settings, "build_sha", ""))
    if build_sha:
        recorded["release.build_sha"] = build_sha
    # A cost identifier, and the only one this side owns: every SQL statement the
    # run executes is billed to this warehouse, and a trace that does not name it
    # cannot be joined to `system.billing.usage`.
    warehouse = _text(getattr(settings, "warehouse_id", ""))
    if warehouse:
        recorded["deployment.warehouse_id"] = warehouse
    return recorded


def _text(value: Any) -> str:
    return value.strip() if isinstance(value, str) else ""

File: player-insights-agent/shared/correlation.ts
/**
 * The one id that names a question everywhere it is recorded.
 *
 * Six systems record something about a single ask -- the app's own log, the run
 * ledger in Lakebase, the model serving span, the Genie conversation, the Vector
 * Search query, and the MLflow trace -- and until this existed each named it
 * something different. Diagnosing one slow or wrong answer meant matching rows
 * by timestamp across all six, which works until two people ask at once.
 *
 * WHY THE BROWSER MINTS IT AND NOT THE SERVER. The failure that most needs a
 * correlation id is the one where the server never answers: a release replaces
 * the app mid-question and `fetch` rejects with no status, no body and no id
 * (see `AskUnreachable` in client/src/ask-stream.ts). A server-minted id exists
 * only in a response that never arrived. An id minted before the request leaves
 * is in the browser's hands whatever happens to the server, and the server
 * logged it on the way in, so the two can still be joined afterwards.
 *
 * WHY IT IS NOT THE RUN'S PRIMARY KEY, WHICH IS THE IMPORTANT PART. The run
 * ledger's `run_id` stays minted by the server. A caller-supplied value used as
 * a primary key lets a caller name a row: `createOrGetRun` inserts on
 * `run_id`, so a client replaying somebody else's id conflicts with their row,
 * and the read that follows is scoped to the caller's own email and finds
 * nothing -- three retries and a self-inflicted denial. That is not a data leak,
 * and it is still a caller reaching a table it has no business addressing. So
 * the two ids are deliberately separate: `run_id` is a database key, and this is
 * the identifier every human-facing and cross-system record carries. When the
 * browser sends nothing usable they hold the same value, which is what the app
 * did before this module and why nothing regresses without it.
 */

/**
 * The header the browser sends it on.
 *
 * `x-` prefixed and app-specific rather than one of the tracing conventions
 * (`traceparent`, `X-Request-ID`). Databricks Apps sits behind a proxy that
 * sets and rewrites headers of its own, and a name that collides with one it
 * manages is a name that arrives changed or not at all.
 */
export const CORRELATION_HEADER = 'x-pia-correlation-id';

/** Distinguishes ours in a log line that also carries platform ids. */
export const CORRELATION_PREFIX = 'req-';

/**
 * A hyphenated lowercase UUID, and nothing else, after the prefix.
 *
 * STRICT BECAUSE THE VALUE IS PRINTED AND STORED. It goes into server log lines,
 * into a Lakebase column, into a trace attribute and back to the browser in a
 * refusal body. A caller-controlled string in a log line can forge a line; one
 * in a trace attribute can carry a question's text back out of a system that
 * deliberately does not record questions. A UUID shape can do neither, and a
 * client that cannot produce one gets a server-minted id rather than an error --
 * there is nothing here worth failing a question over.
 */
const CORRELATION_SHAPE = /^req-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/** A fresh id. The same function on both sides of the wire, deliberately. */
export function mintCorrelationId(): string {
  return `${CORRELATION_PREFIX}${crypto.randomUUID()}`;
}

/**
 * The caller's id if it is one we will print, otherwise null.
 *
 * Null is ordinary: a `curl`, an older browser build, and the benchmark runner
 * all send nothing. The caller is never told, because a rejected correlation id
 * changes nothing about whether their question can be answered.
 */
export function usableCorrelationId(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  const trimmed = value.trim();
  return CORRELATION_SHAPE.test(trimmed) ? trimmed : null;
}
